### PR TITLE
fix(Pilot Sheet): Refactor Pilot groups for performance & roster improvements

### LINKED ISSUE #1611: Pilot Group Reordering
The ability to reorder groups of pilots on comp/con useful for both players and gm's 

Low Priority but it would be nice to have especially for people that manage many groups of pilots.

### LINKED ISSUE #1709: [BUG] Pilot Groups hidden on login
<strong>Submitted At: </strong> 10/9/2021 11:05:40<br><br><strong>Browser: </strong> Firefox<br><strong>Platform: </strong> Desktop<br><br><strong>Issue Description: </strong><br> Pilot Groups do not initialize upon logging into CompCon.<br><br><strong>Reproduction Steps:<br> </strong> 1. Be logged into CompCon, create a pilot, put them in a custom-named group, and sync the pilot
2. Open CompCon with no local storage (e.g. in Incognito mode)
3. Log into account from step 1
4. Navigate to Pilot Roster
5. Observe that the custom Pilot Group from step 1 is not present
6. Refresh page
7. Observe that the custom Pilot Group from step 1 is present<br><br><strong>Expected Result:<br> </strong> Upon log in, pilot groups are initialized in the app to prevent confusion and/or necessitating a page reload.<br><br><strong>Additional Context:<br> </strong> This group initialization was previously added for importing pilots, so similar code added to the "login startup" may solve this issue.
<br><br><br><strong>Stack Trace: </strong> 

**Discord Handle:**
Valkyrion#6196<br>

### LINKED ISSUE #1713: [BUG] Pilot Roster Performance suffers for large pilot counts
**Describe the bug**
When dealing with large amounts of pilots (~100), pilot roster performance becomes sluggish.  The click-and-drag reordering becomes inconsistent and unresponsive at times.  When using Chrome, the browser sometimes runs out of memory in egregious cases.

**To Reproduce**
1. Open CompCon on Firefox
2. Create ~100 pilots (user-reported number was about 150)
3. Attempt to use the click-and-drag reordering
4. Observe the inconsistency when handling click and drag

**Expected behavior**
The Pilot Roster maintains serviceable performance for large pilot counts.

**Screenshots**
Videos provided by a PilotNet User.  First one is Firefox on Windows 10, ~150 pilots:
https://gyazo.com/aac44f45d919eef94d808792e5908d34

Second is Chrome on Windows 10, ~150 pilots:
https://i.gyazo.com/2c73dd87fcc6a9753f4d66c18cae674f.mp4

**Desktop (please complete the following information):**
 - OS: Windows 10
 - Browser Firefox
 - Version 93

**Additional context**
On Windows 10 + Chrome, the "ghost" of the pilot banner being dragged appears several entries above where the pilot banner originally displayed, regardless of number of pilots.  Unsure if this is a "top of screen" or a "misaligned index" issue, but the ghost basically is offset quite a bit from the actual mouse position.  If pilot click-and-drag is getting redone, this could maybe be addressed as a bonus objective.

**Contact**
Valkyrion#6196

### PR DESCRIPTION
**Pull request by msprijatelj**
# Description
~~This is an initial commit for improving performance in the Pilot Roster.  It lays the groundwork for using nested Draggable components, utilizing the yet-unused `pilot_groups.json` for persisting Roster organization, and fixes autoscrolling issues (and drag-and-drop ghost issues in Chrome).  It is far from complete; PilotStore needs cleaned up to fit the new Group paradigm, the Pilot Groups object needs to be used directly in Draggable, and it should probably be load-tested to see if it can handle large lists of pilots.~~

UPDATE: 
This change refactors the way we organize the pilot roster, storing group configuration in `pilot_groups.json`.  This file tracks the Group Name, Hidden status, and list of Pilot IDs of pilots in the group.  The groups themselves have become draggable, and their Hidden state is now persistent between visits to the Pilot Roster.

Current challenges: 
- How to meaningfully load test performance for high pilot counts.
- When performing a Cloud Pull on a freshly-deleted CompCon, there is no guarantee that the groups will retain their previous order, nor pilots within a group.  This could probably be solved by concatenating the `pilotIDs` fields of all the groups in order, mapping them to their Pilot, and then setting `this.Pilots` to the resultant map.  However, it's not immediately clear when this pilot reassignment should occur, and how reassigning `this.Pilots` on a regular basis would impact performance.
- ~~EDIT: One more snag: The draggable doesn't seem to be working for the Large Tiles viewing mode, though it's working fine for List and Small Tiles.  In Large Tiles, the tile just drags to the last spot in the list, refusing to move elsewhere.~~ EDIT 2: This has been remedied; see comments below.

Regardless of these current challenges, I feel that this PR provides enough value to ship as-is. Please let me know if you have any questions or feedback!

## Issue Number
Closes #1713 (hopefully, monitor this one)
Closes #1709
Closes #1611

## Type of change
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

### CODE REVIEW

**Comment by msprijatelj** [comment]
Lessons in VueDraggables learned from this:
- When working with a draggable, make sure that the object referenced by the `v-model` or `list` attributes is the *actual object you want to sort*.  
- Make sure the `v-for` component references the same object as the draggable's `v-model`/`list`.
- Sometimes, the browser won't play nice with the draggable, preventing its innate autoscrolling and causing other strange issues.  When this happens, use `forceFallback: true`.

File: src/features/pilot_management/Roster/index.vue [hunk 1]
```diff
@@ -48,66 +48,76 @@
...
+        <draggable
+          :list="groups"
+          :disabled="preventDnd"
+          v-bind="dragOptions"
+          @change="pilotStore.moveGroup"
+        >
           <div
-            v-if="!hidden.includes(g)"
-            :style="profile.GetView('roster') !== 'list' ? 'margin-left: -8px; width: 100vw;' : ''"
+            v-for="(g, i) in groups"
```
Comment (msprijatelj) [inline src/features/pilot_management/Roster/index.vue:58]: Use `groups` in both the draggable's `list` and the component's `v-for`.

File: src/features/pilot_management/Roster/index.vue [hunk 2]
```diff
@@ -48,66 +48,76 @@
...
+                  @end="dragOff()"
+                  @change="movePilot(g.name, $event)"
+                >
+                  <component
+                    :is="pilotCardType"
+                    v-for="(id, j) in g.pilotIDs"
+                    :key="`${pilotCardType}_${j}`"
+                    :pilot="getPilotFromId(id)"
+                    :small="profile.GetView('roster') === 'small-cards'"
+                    :dragging="drag"
```
Comment (msprijatelj) [inline src/features/pilot_management/Roster/index.vue:114]: When `forceFallback` was set to `true` in the dragOptions to fix Autoscrolling and a weird offset Ghost icon in Chrome, a new error emerged in Firefox: Whenever the user releases the dragged object, CompCon would treat it like a released click and navigate to the Pilot Sheet.  Had to finagle a short timeout function `dragOff` to keep it from propagating.  If you have recommendations on how to better solve this, please let me know.  I tried `click.stop` and `click.prevent` with no luck, but maybe I was using them the wrong way. :shrug:

**Comment by msprijatelj** [comment]
Regarding the Large Tiles issue: I'm finding that the large tiles drag properly when `forceFallback` is disabled, but this causes issues with autoscrolling and some graphical issues in Chrome.  I'll see if I can't troubleshoot it further.

**Review by TranslucentSabre — APPROVED** [review]

**Pull request merged**